src/ffi2/datetime.rs

```diff
@@ -0,0 +1,354 @@
+    pub seconds: c_int,
+    pub microseconds: c_int,
+}
+
+// C API Capsule
+unsafe impl Sync for PyDateTime_CAPI {}
+
+lazy_static! {
+    pub static ref PyDateTimeAPI: PyDateTime_CAPI = unsafe { PyDateTime_IMPORT() };
+}
+
+#[inline(always)]
+pub unsafe fn PyDateTime_IMPORT() -> PyDateTime_CAPI {
+    // PyDateTime_CAPSULE_NAME is a macro in C
+    let PyDateTime_CAPSULE_NAME = CString::new("datetime.datetime_CAPI").unwrap();
+
+    let capsule = PyCapsule_Import(PyDateTime_CAPSULE_NAME.as_ptr(), 0);
+    *(capsule as *const PyDateTime_CAPI)
+}
+
+//
+// Type Check macros
+//
+#[inline(always)]
+pub unsafe fn PyDate_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.DateType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDate_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.DateType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDateTime_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.DateTimeType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDateTime_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.DateTimeType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyTime_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.TimeType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyTime_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.TimeType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDelta_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.DeltaType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyDelta_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.DeltaType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyTZInfo_Check(op: *mut PyObject) -> c_int {
+    PyObject_TypeCheck(op, PyDateTimeAPI.TZInfoType) as c_int
+}
+
+#[inline(always)]
+pub unsafe fn PyTZInfo_CheckExact(op: *mut PyObject) -> c_int {
+    (Py_TYPE(op) == PyDateTimeAPI.TZInfoType) as c_int
+}
+
+//
+// Accessor functions
+//
+macro_rules! _access_field {
+    ($obj:expr, $type: ident, $field:tt) => {
+        (*($obj as *mut $type)).$field
+    };
+}
+
+// Accessor functions for PyDateTime_Date
+// Note: These have nonsensical names
+#[macro_export]
+macro_rules! PyDateTime_GET_YEAR {
+    // This is a macro in the C API and it's difficult to get the same behavior
+    // without making it a macro in Rust as well, or playing with pointers
+    ($o: expr) => {
+        (((*$o).data[0] as c_int) << 8) | ((*$o).data[1] as c_int)
+    };
+}
+
+#[inline(always)]
+pub unsafe fn PyDateTime_Date_GET_YEAR(o: *mut PyObject) -> c_int {
+    PyDateTime_GET_YEAR!(o as *mut PyDateTime_Date)
+}
+
+#[inline(always)]
+pub unsafe fn PyDateTime_DateTime_GET_YEAR(o: *mut PyObject) -> c_int {
+    PyDateTime_GET_YEAR!(o as *mut PyDateTime_DateTime)
+}
+
+#[macro_export]
+macro_rules! PyDateTime_GET_MONTH {
+    ($o: expr) => {
+        (*$o).data[2] as c_int
```

Comment:
Clippy complains about this:

```
warning: casting u8 to i32 may become silently lossy if types change
   --> src/ffi3/datetime.rs:271:9
    |
271 |         (*$o).data[3] as c_int
    |         ^^^^^^^^^^^^^^^^^^^^^^ help: try: `i32::from((*$o).data[3])`
...
277 |     PyDateTime_GET_DAY!(o as *mut PyDateTime_Date)
    |     ---------------------------------------------- in this macro invocation
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#cast_lossless
```

Reply:
Not sure I see any *good* way around it. We could store `month` and `day` as `u8` in the Rust API I suppose, though we'll still have this problem for `year`.

Hopefully the test suite will alert us if this becomes a problem.

Reply:
`c_int::from((*$o).data[3]) works and makes clippy happy, so I'd take that